### PR TITLE
Implement blueprint.

### PR DESCRIPTION
**Pull request by dmtucker**
This is nearly identical to `host list`.

### CODE REVIEW

File: yaas/blueprint.py [hunk 1]
```diff
@@ -1,4 +1,51 @@
...
+    """ List all blueprints stored on the Ambari server. """
+    response = requests.get(
+        config.href('/api/v1/blueprints'),
+        **config.requests_opts())
+    if config.args.raw:
+        pprint.pprint(response.json())
+        sys.exit(0)
+    try:
+        response.raise_for_status()
+    except requests.exceptions.HTTPError:
```
Comment (karth295) [inline yaas/blueprint.py]: Rebase against other PR's?

File: yaas/blueprint.py [hunk 2]
```diff
@@ -1,4 +1,138 @@
...
+    subcommands[subargs.subcommand](
+        subparsers.choices[subargs.subcommand],
+        extra)
+
+
+def _add(parser, args):
+    """ Add a blueprint to the Ambari server. """
+    parser.add_argument(
+        '--validate-topology',
+        default='true',
```
Comment (dmtucker) [inline yaas/blueprint.py]: not necessary... Without it, Ambari's default is used which would be better.

Comment (karth295) [inline yaas/blueprint.py]: Shouldn't that be a boolean:

http://stackoverflow.com/questions/15008758/parsing-boolean-values-with-argparse

Explicit is better than implicit--I like having the default in the code.

File: yaas/blueprint.py [hunk 3]
```diff
@@ -1,4 +1,138 @@
...
+    for blueprint_file in subargs.blueprint:
+        with open(blueprint_file, 'r') as f:
+            blueprints.append(f.read())
+    for blueprint in blueprints:
+        blueprint = json.loads(blueprint)
+        response = requests.post(
+            config.href(
+                '/api/v1/blueprints/{name}'.format(
+                    name=blueprint['Blueprints']['blueprint_name'])),
+            data=json.dumps(blueprint),  # `json=blueprint` would be better, but it doesn't work!
```
Comment (dmtucker) [inline yaas/blueprint.py:53]: Is this bugworthy upstream?

Comment (karth295) [inline yaas/blueprint.py:53]: What version of requests are you using?

http://stackoverflow.com/questions/9733638/post-json-using-python-requests

File: yaas/blueprint.py (hunk 2 above)
Comment (dmtucker) [inline yaas/blueprint.py]: - Not imposing `bool` here future-proofs better (e.g. Maybe `validate_topology=configurations_only` could be valid in the future).
- `argparse`+`bool` is a pain in the ass.
- The default is defined upstream which means if we are explicit here, a code change will be required if that default ever changes.
- Using `bool` requires us to solve a problem Ambari already solves: accepting `false` v. `False`, etc.

File: yaas/blueprint.py (hunk 3 above)
Comment (dmtucker) [inline yaas/blueprint.py:53]: Python 2.7.10
Requests 2.5.0

This is the weird content-type thing I mentioned before.

File: yaas/blueprint.py (hunk 2 above)
Comment (karth295) [inline yaas/blueprint.py]: Ah, good points.

File: yaas/blueprint.py [hunk 4]
```diff
@@ -1,4 +1,138 @@
...
+    """ Add a blueprint to the Ambari server. """
+    parser.add_argument(
+        '--validate-topology',
+        default='true',
+        help="Specify whether to validate a blueprint on upload.")
+    parser.add_argument(
+        'blueprint',
+        nargs='*',
+        help="Specify a blueprint to add to the Ambari server.")
+    subargs, extra = parser.parse_known_args(args)
```
Comment (dmtucker) [inline yaas/blueprint.py]: We should probably use `.parse_args(args)` at the final subparser.

File: yaas/blueprint.py (hunk 3 above)
Comment (dmtucker) [inline yaas/blueprint.py:53]: https://issues.apache.org/jira/browse/AMBARI-9016

**Pull request merged**